web/server/codechecker_server/task_executors/abstract_task.py

```diff
@@ -0,0 +1,183 @@
+#  LLVM Exceptions. See LICENSE for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# -------------------------------------------------------------------------
+"""
+Contains the base class to be inherited and implemented by all background task
+types.
+"""
+import os
+import pathlib
+import shutil
+from typing import Optional
+
+from codechecker_common.logger import get_logger
+
+from ..database.config_db_model import BackgroundTask as DBTask
+
+
+LOG = get_logger("server")
+
+
+class TaskCancelHonoured(Exception):
+    """
+    Specialised tag exception raised by `AbstractTask` implementations in a
+    checkpoint after having checked that their ``cancel_flag`` was set, in
+    order to terminate task-specific execution and to register the
+    cancellation's success by the `AbstractTask.execute` method.
+
+    This exception should **NOT** be caught by user code.
+    """
+
+    def __init__(self, task_obj: "AbstractTask"):
+        super().__init__(f"Task '{task_obj.token}' honoured CANCEL request.")
+        self.task_obj = task_obj
+
+
+class AbstractTask:
+    """
+    Base class implementing common execution and bookkeeping methods to
+    facilitate the dispatch of tasks to background worker processes.
+
+    Instances of this class **MUST** be marshallable by ``pickle``, as they
+    are transported over an IPC `Queue`.
+    It is important that instances do not grow too large, as the underlying
+    OS-level primitives of a `Queue` can get full, which can result in a
+    deadlock situation.
+
+    The run-time contents of the instance should only contain the bare minimum
+    metadata required for the implementation to execute in the background.
+
+    Implementors of subclasses **MAY REASONABLY ASSUME** that an
+    `AbstractTask` scheduled in the API handler process of a server will be
+    actually executed by a background worker in the same process group, on the
+    same machine instance.
+    """
+
+    def __init__(self, token: str, data_path: Optional[pathlib.Path]):
+        self._token = token
+        self._data_path = data_path
+
+    @property
+    def token(self) -> str:
+        """Returns the task's identifying token, its primary ID."""
+        return self._token
+
+    @property
+    def data_path(self) -> Optional[pathlib.Path]:
+        """
+        Returns the filesystem path where the task's input data is prepared.
+        """
+        return self._data_path
+
+    def destroy_data(self):
+        """
+        Deletes the contents of `data_path`.
+        """
+        if not self._data_path:
+            return
+
+        try:
+            shutil.rmtree(self._data_path)
+        except Exception as ex:
+            LOG.warning("Failed to remove background task's data_dir at "
+                        "'%s':\n%s", self.data_path, str(ex))
+
+    def _implementation(self, _task_manager: "TaskManager") -> None:
```

Comment:
Shouldn't it get an `@abstractmethod` annotation?

Reply:
Unfortunately, `@abstractmethod` isn't a built-in, but coming from the `abc` library, which makes the object not pickleable.

Also, methods that raise `NotImplementedError()` are type-checked by PyRight and there is a linter warning in the editor that your class is not proper.